src/transformers/generation/watermarking.py

```diff
@@ -0,0 +1,232 @@
+                target = ngram_example[-1]
+                ngram_to_watermark_lookup[ngram_example] = self._get_ngram_score_cached(prefix, target)
+
+            if self.ignore_repeated_ngrams:
+                # counts a green/red hit once per unique ngram.
+                # num total tokens scored becomes the number unique ngrams.
+                num_tokens_scored_batch[batch_idx] = len(frequencies_table.keys())
+                green_token_count_batch[batch_idx] = sum(ngram_to_watermark_lookup.values())
+            else:
+                num_tokens_scored_batch[batch_idx] = sum(frequencies_table.values())
+                green_token_count_batch[batch_idx] = sum(
+                    freq * outcome
+                    for freq, outcome in zip(frequencies_table.values(), ngram_to_watermark_lookup.values())
+                )
+        return num_tokens_scored_batch, green_token_count_batch
+
+    def _compute_z_score(self, green_token_count: np.array, total_num_tokens: np.array) -> np.array:
+        expected_count = self.greenlist_ratio
+        numer = green_token_count - expected_count * total_num_tokens
+        denom = np.sqrt(total_num_tokens * expected_count * (1 - expected_count))
+        z = numer / denom
+        return z
+
+    def _compute_pval(self, x, loc=0, scale=1):
+        z = (x - loc) / scale
+        return 1 - (0.5 * (1 + np.sign(z) * (1 - np.exp(-2 * z**2 / np.pi))))
+
+    def __call__(
+        self,
+        input_ids: torch.LongTensor,
+        z_threshold: float = 3.0,
+        return_prediction: bool = True,
+        return_dict: bool = False,
+    ) -> Union[WatermarkDetectorOutput, np.array]:
+        """
+        Args:
+        input_ids (`torch.LongTensor`):
+            The watermark generated text. It is advised to remove the prompt, which can affect the detection.
+        z_threshold (`Dict`, *optional*):
+            Changing this threshold will change the sensitivity of the detector.
+        return_prediction (`bool`):
+            Whether to return user-friendly predictions as a boolean array or not.
+        return_dict (`bool`):
+            Whether to return `~generation.WatermarkDetectorOutput` or not. If not it will return `z_score` if
+            `return_prediction=False` else `prediction`,
+
+        Return:
+            [`~generation.WatermarkDetectorOutput`] or `np.array`: A [`~utils.ModelOutput`] (if `return_dict=True`
+            otherwise a `np.array`.
+
+        """
+
+        if input_ids.shape[-1] - self.processor.context_width < 1:
+            raise ValueError(
+                f"Must have at least `1` token to score after the first "
+                f"min_prefix_len={self.processor.context_width} tokens required by the seeding scheme."
+            )
+        # Let;s assume that if one batch start with `bos`, all batched also do
+        if input_ids[0, 0] == self.bos_token_id:
+            input_ids = input_ids[:, 1:]
```

Comment:
Two notes:
1. In some encoder-decoder models, the first token may be defined by another flag (`decoder_start_token_id`), which we should also check.
2. The `context_width` check should be done after this cropping

Reply:
1. The `bos` here is the token passed by a user when initializing. I can add somewhere in docs that they should pass `decoder_start_token_id` or `bos`, whichever they have in config.
2. yep

Reply:
>  I can add somewhere in docs that they should pass decoder_start_token_id or bos, whichever they have in config.

Uhmm that could add some friction (the user needs to know those details). How about we pass a `generation_config`, and we handle the logic in the detector?

Reply:
Hmm, yeah, seems like it's getting hard to init the detector. I think we can ask for the model config then, and get "vocab size", "bos token id" directly from config. 

It would be more comfortable to pass in the tokenizer, but I cannot see any "decoder start token id" in it  :(

Reply:
> but I cannot see any "decoder start token id" in it 

I believe this is a generation-time attribute, for models that support several start tokens